Add documentation about .gdextension file

paddy-exe opened the pull request:
This is a PR in the effort to improve the GDExtension docs for users

Adds details how to configure the `.gdextension` file and aids as a quick reference guide to add support for platforms

dsnopek commented on tutorials/scripting/gdextension/gdextension_file.rst:
```diff
@@ -0,0 +1,95 @@
+Introduction
+------------
+
+The ``.gdextension`` file in your project contains the instructions for the editor how to parse
+the GDExtensions in projects. The instructions are separated into specific sections. This page
```
```suggestion
The ``.gdextension`` file in your project contains the instructions for how to load
the GDExtension. The instructions are separated into specific sections. This page
```

dsnopek commented on tutorials/scripting/gdextension/gdextension_file.rst:
```diff
@@ -0,0 +1,95 @@
++-------------------------------+------------------------------------------------------------------------------------------------------+
+| Instruction                   | Description                                                                                          |
++===============================+======================================================================================================+
+| **entry_smbol**               | Name of the entry function of the loaded GDExtension. This function should be defined in             |
```
```suggestion
| **entry_smbol**               | Name of the entry function for initializing the GDExtension. This function should be defined in             |
```

dsnopek commented on tutorials/scripting/gdextension/gdextension_file.rst:
```diff
@@ -0,0 +1,95 @@
+| **entry_smbol**               | Name of the entry function of the loaded GDExtension. This function should be defined in             |
+|                               | the ``register_types.cpp`` file. Adding this is necessary for the extension to work.                 |
++-------------------------------+------------------------------------------------------------------------------------------------------+
+| **compatability_minimum**     | Minimum compatible version. This prevents older editor binaries from loading the extension.          |
```
```suggestion
| **compatability_minimum**     | Minimum compatible version. This prevents older versions of Godot from loading extensions that depend on features from newer versions of Godot.          |
```

dsnopek commented on tutorials/scripting/gdextension/gdextension_file.rst:
```diff
@@ -0,0 +1,95 @@
+Libraries section
+-----------------
+
+In this section you can set the paths to your compiled binaries of your GDExtension libraries.
```
```suggestion
In this section you can set the paths to the compiled binaries of your GDExtension libraries.
```

dsnopek commented on tutorials/scripting/gdextension/gdextension_file.rst:
```diff
@@ -0,0 +1,95 @@
+
+In this section you can set the paths to your compiled binaries of your GDExtension libraries.
+By setting feature flags you can filter which version should be loaded and exported with your
+game depending on which feature flags are active.
```
I think we should add something about how all feature flags on each line must match (so, `macos.debug` means that it will be loaded if Godot has both "macos" and "debug" feature flags), and that each line is evaluated in order from top-to-bottom.

dsnopek reviewed (commented):
Thanks, this looks great!

Only a few notes, mostly small wording changes for clarity.

aaronfranke commented on tutorials/scripting/gdextension/gdextension_file.rst:
```diff
@@ -0,0 +1,95 @@
+^^^^^^^^^^^^
+
++-------------------------------+------------------------------------------------------------------------------------------------------+
+| Flag                          | Description                                                                                          |
++===============================+======================================================================================================+
+| **x86_64**                    | 64-bit Windows                                                                                       |
++-------------------------------+------------------------------------------------------------------------------------------------------+
+| **x86_32**                    | 32-bit Windows                                                                                       |
```
The flags are architecture-specific, not Windows-specific. x86_64 means 64-bit x86, not 64-bit Windows.

```suggestion
| Flag                          | Description                                                                                          |
+===============================+======================================================================================================+
| **x86_64**                    | 64-bit x86                                                                                           |
+-------------------------------+------------------------------------------------------------------------------------------------------+
| **x86_32**                    | 32-bit x86                                                                                           |
+-------------------------------+------------------------------------------------------------------------------------------------------+
| **arm32**                     | 32-bit Arm                                                                                           |
+-------------------------------+------------------------------------------------------------------------------------------------------+
| **arm64**                     | 64-bit Arm                                                                                           |
+-------------------------------+------------------------------------------------------------------------------------------------------+
| **rv64**                      | 64-bit RISC-V                                                                                        |
+-------------------------------+------------------------------------------------------------------------------------------------------+
| **ppc32**                     | 32-bit PowerPC                                                                                       |
+-------------------------------+------------------------------------------------------------------------------------------------------+
| **ppc64**                     | 64-bit PowerPC                                                                                       |
+-------------------------------+------------------------------------------------------------------------------------------------------+
| **wasm32**                    | 32-bit WebAssembly                                                                                   |
```

paddy-exe replied:
Noticed already. Thank you!

aaronfranke commented on tutorials/scripting/gdextension/gdextension_file.rst:
```diff
@@ -0,0 +1,136 @@
+| **entry_smbol**               | Name of the entry function for initializing the GDExtension. This function should be defined in      |
+|                               | the ``register_types.cpp`` file. Adding this is necessary for the extension to work.                 |
++-------------------------------+------------------------------------------------------------------------------------------------------+
+| **compatability_minimum**     | Minimum compatible version. This prevents older versions of Godot from loading extensions that       |
+|                               | depend on features from newer versions of Godot. **Only supported for Godot 4.1+**                   |
++-------------------------------+------------------------------------------------------------------------------------------------------+
+| **compatability_maximum**     | Minimum compatible version. This prevents older versions of Godot from loading extensions that       |
+|                               | depend on features from newer versions of Godot. **Only supported for Godot 4.3+**                   |
++-------------------------------+------------------------------------------------------------------------------------------------------+
+| **compatability_minimum**     | Minimum compatible version. This prevents older versions of Godot from loading extensions that       |
+|                               | depend on features from newer versions of Godot.                                                     |
```
Duplicate `compatability_minimum`

aaronfranke commented on tutorials/scripting/gdextension/gdextension_file.rst:
```diff
@@ -0,0 +1,136 @@
++-------------------------------+------------------------------------------------------------------------------------------------------+
+| **single**                    | single-precision build                                                                               |
++-------------------------------+------------------------------------------------------------------------------------------------------+
+| **64**                        | 64-bit build (any architecture)                                                                      |
++-------------------------------+------------------------------------------------------------------------------------------------------+
+| **32**                        | 32-bit build (any architecture)                                                                      |
```
Given that, for example, x86_64 and arm64 are not binary-compatible, I don't think it's worth mentioning/encouraging `64` or `32` as options. I think we should link to Godot's feature tag documentation for a complete list, and on this documentation page only list the ones we encourage using for GDExtension. Same with `x86` and `arm`, those should probably not be used with GDExtension, instead `arm64` etc should be used.

Also, Godot doesn't support 64-bit WebAssembly yet.

paddy-exe replied:
I removed the feature tags that should not be advertised as prominently as suggested

RedMser commented on tutorials/scripting/gdextension/gdextension_file.rst:
```diff
@@ -0,0 +1,123 @@
++-------------------------------+------------------------------------------------------------------------------------------------------+
+| Instruction                   | Description                                                                                          |
++===============================+======================================================================================================+
+| **entry_smbol**               | Name of the entry function for initializing the GDExtension. This function should be defined in      |
```
```suggestion
| **entry_symbol**               | Name of the entry function for initializing the GDExtension. This function should be defined in      |
```

AThousandShips replied:
```suggestion
| **entry_symbol**              | Name of the entry function for initializing the GDExtension. This function should be defined in      |
```
Rather

RedMser commented on tutorials/scripting/gdextension/gdextension_file.rst:
```diff
@@ -0,0 +1,123 @@
+| **entry_smbol**               | Name of the entry function for initializing the GDExtension. This function should be defined in      |
+|                               | the ``register_types.cpp`` file. Adding this is necessary for the extension to work.                 |
++-------------------------------+------------------------------------------------------------------------------------------------------+
+| **compatability_minimum**     | Minimum compatible version. This prevents older versions of Godot from loading extensions that       |
```
```suggestion
| **compatibility_minimum**     | Minimum compatible version. This prevents older versions of Godot from loading extensions that       |
```

RedMser commented on tutorials/scripting/gdextension/gdextension_file.rst:
```diff
@@ -0,0 +1,123 @@
+| **compatability_minimum**     | Minimum compatible version. This prevents older versions of Godot from loading extensions that       |
+|                               | depend on features from newer versions of Godot. **Only supported for Godot 4.1+**                   |
++-------------------------------+------------------------------------------------------------------------------------------------------+
+| **compatability_maximum**     | Maximum compatible version. This prevents newer versions of Godot from loading the extension.        |
```
```suggestion
| **compatibility_maximum**     | Maximum compatible version. This prevents newer versions of Godot from loading the extension.        |
```

paddy-exe commented:
Thanks for the feedback! Should be good now

AThousandShips commented on tutorials/scripting/gdextension/gdextension_file.rst:
```diff
@@ -0,0 +1,123 @@
+By setting feature flags you can filter which version should be loaded and exported with your
+game depending on which feature flags are active. Every feature flag must match to Godot's
+feature flags. For instance ``macos.debug`` means that it will be loaded if Godot has both the
+``macos`` and ``debug`` flag active. Each line of the section is evaluated from top to bottom.
```
Does this not support custom feature flags on export? This and below implies only the official ones are accepted

paddy-exe commented:
Updated to feedback, added another sections about icons support and removed the icons section from the Example page as it makes no sense to have it double

dsnopek commented on tutorials/scripting/gdextension/gdextension_file.rst:
```diff
@@ -0,0 +1,141 @@
+| **compatibility_maximum**     | Maximum compatible version. This prevents newer versions of Godot from loading the extension.        |
+|                               | **Only supported for Godot 4.3+**                                                                    |
++-------------------------------+------------------------------------------------------------------------------------------------------+
+| **reloadable**                | Reloads the extension upon recompilation. Reloading is not supported for every exposed class.        |
+|                               | This flag should be mainly used for developing or debugging an extension.                            |
```
What is meant by "Reloading is not supported for every exposed class"? It should work for any class, although, some can have additional considerations, for example, editor plugins.

It may be worth pointing out that not all GDExtension bindings support it? We could mention that godot-cpp for Godot 4.2+ does - I'm not sure it makes sense trying to list other bindings, because that list will always be growing

paddy-exe replied:
I wasn't sure if it is already supported for every class especially the editor plugins classes due to how they are handled with reloading. If are you certain then I will change it. I would only list the official godot-cpp support as we other language bindings can say themselves if they support reload support or not

paddy-exe replied:
Updated now. Let me know what you think

dsnopek replied:
Looks good to me!

dsnopek commented on tutorials/scripting/gdextension/gdextension_file.rst:
```diff
@@ -0,0 +1,141 @@
+| **reloadable**                | Reloads the extension upon recompilation. Reloading is not supported for every exposed class.        |
+|                               | This flag should be mainly used for developing or debugging an extension.                            |
++-------------------------------+------------------------------------------------------------------------------------------------------+
+| **android_aar_plugin**        | The GDExtension is part of a :ref:`v2 Android plugin <doc_android_plugin>`. During export this flag  |
```
It's not clear from this description that this is a boolean property. Perhaps we should be listing the type for each property?

dsnopek replied:
Thanks!

dsnopek commented on tutorials/scripting/gdextension/gdextension_file.rst:
```diff
@@ -0,0 +1,141 @@
+---------------------
+
++-------------------------------+------------------------------------------------------------------------------------------------------+
+| Instruction                   | Description                                                                                          |
```
I'm not sure about calling these "Instruction"s. Maybe "Property" instead?

```suggestion
| Property                  | Description                                                                                          |
```

dsnopek commented on tutorials/scripting/gdextension/gdextension_file.rst:
```diff
@@ -0,0 +1,141 @@
+-----------------
+
+In this section you can set the paths to the compiled binaries of your GDExtension libraries.
+By setting feature flags you can filter which version should be loaded and exported with your
```
```suggestion
By specifying feature flags you can filter which version should be loaded and exported with your
```

dsnopek commented on tutorials/scripting/gdextension/gdextension_file.rst:
```diff
@@ -0,0 +1,141 @@
+In this section you can set the paths to the compiled binaries of your GDExtension libraries.
+By setting feature flags you can filter which version should be loaded and exported with your
+game depending on which feature flags are active. Every feature flag must match to Godot's
+feature flags or your custom export flags to run in an exported game. For instance ``macos.debug``
```
```suggestion
feature flags or your custom export flags to be loaded in an exported game. For instance ``macos.debug``
```

dsnopek commented on tutorials/scripting/gdextension/gdextension_file.rst:
```diff
@@ -0,0 +1,141 @@
+    linux.release.rv64 = "res://bin/libgdexample.linux.template_release.rv64.so"
+
+
+Here are lists of available built-in options (for more look at the :ref:`feature tags <doc_feature_tags>`):
```
```suggestion
Here are lists of some of the available built-in options (for more look at the :ref:`feature tags <doc_feature_tags>`):
```

dsnopek reviewed (commented):
Thanks! This is looking very nice :-)

I just have a few suggestions, mostly small wording changes for clarity.

aaronfranke commented on tutorials/scripting/gdextension/gdextension_file.rst:
```diff
@@ -0,0 +1,141 @@
+|                               | depend on features from newer versions of Godot. **Only supported for Godot 4.1+**                   |
++-------------------------------+------------------------------------------------------------------------------------------------------+
+| **compatibility_maximum**     | Maximum compatible version. This prevents newer versions of Godot from loading the extension.        |
+|                               | **Only supported for Godot 4.3+**                                                                    |
```
"or later" might also be good to add.

mhilbrunner commented:
@paddy-exe Great work! Merged :) Also thanks to reviewing everyone. Does this apply fully to 4.0/4.1/4.2?

paddy-exe commented:
> @paddy-exe Great work! Merged :) Also thanks to reviewing everyone. Does this apply fully to 4.0/4.1/4.2?

Thanks! :) This should apply for 4.0, 4.1 and 4.2 as well, yes.

mhilbrunner commented:
Cherry-picked to 4.2 in https://github.com/godotengine/godot-docs/pull/9648.
